# Aging Curves for Top Players

The analysis code lives in the `mlb_aging` package, not in this notebook. Run
`uv sync --all-extras` once, then point this notebook's kernel at the project
venv. Feature engineering, model fitting, curve tracing and scoring are all
imported; what stays here is the narrative and the plots.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from mlb_aging.dataset import build_training_frame, load_data
from mlb_aging.metrics import ELITE_WAR_THRESHOLD, ELITE_WRC_THRESHOLD, get_metric
from mlb_aging.pipeline import run_metric
from mlb_aging.plots import plot_aging_curve

WRC = get_metric("wRC+")
WAR = get_metric("WAR")
DEF = get_metric("Def")

### EDA

In [2]:
# How many players clear each elite threshold?
fig, axes = plt.subplots(1, 2, figsize=(14, 4))

for ax, spec, bins in (
    (axes[0], WRC, [-np.inf] + list(range(50, 160, 10)) + [np.inf]),
    (axes[1], WAR, [-np.inf] + list(np.arange(0.0, 6.0, 0.5)) + [np.inf]),
):
    train_data, _ = load_data(spec)
    frame = build_training_frame(train_data, spec)
    careers = frame.groupby("IDfg")[spec.career_mean_col].first()
    sns.countplot(x=pd.cut(careers, bins=bins), ax=ax)
    ax.set_title(f"career {spec.name} per player")
    ax.tick_params(axis="x", rotation=90)

plt.tight_layout()
plt.show()

/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/2992320265.py:16: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


### Original wRC+ Aging Curve

In [3]:
# Trained on everyone, then scored twice: on the whole test set, and on the
# elite subset only. Same model, two populations.
wrc_all = run_metric(WRC)
wrc_all_on_elite = run_metric(WRC, test_threshold=ELITE_WRC_THRESHOLD)

print(wrc_all.summary())
print(f"  scored on career wRC+ > {ELITE_WRC_THRESHOLD} only: "
      f"MAE {wrc_all_on_elite.test_mae:.4f}  (n_test={wrc_all_on_elite.n_test})")

plot_aging_curve(wrc_all.curve, WRC)
plt.show()

wRC+: peak age 26.99  |  peak value 102.2726  |  test MAE 19.7730  (n_train=13636, n_test=887)
  scored on career wRC+ > 110 only: MAE 20.5633  (n_test=411)


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/1539077103.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


### Original WAR Aging Curve

In [4]:
war_all = run_metric(WAR)
war_all_on_elite = run_metric(WAR, test_threshold=ELITE_WAR_THRESHOLD)

print(war_all.summary())
print(f"  scored on career WAR > {ELITE_WAR_THRESHOLD} only: "
      f"MAE {war_all_on_elite.test_mae:.4f}  (n_test={war_all_on_elite.n_test})")

plot_aging_curve(war_all.curve, WAR)
plt.show()

WAR: peak age 27.01  |  peak value 1.7081  |  test MAE 1.4970  (n_train=13636, n_test=887)
  scored PA-weighted; G-weighted (matching the fit) would be 1.4719
  scored on career WAR > 2.5 only: MAE 1.7480  (n_test=376)


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/2623329855.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


The all-player models are less accurate when evaluated specifically on top performers. The wRC+
model achieves a test MAE of 19.77 on the full test set but 20.56 when restricted to strong
players; the WAR model's test MAE rises from 1.497 (all players) to 1.748 (strong players).
This suggests that the aging dynamics of elite players may differ from the average player the
models were trained on.

The following sections train specialist models restricted to top players using two filters:
career wRC+ ≥ 110 for hitting ability and career WAR ≥ 2.5 for overall value. Each filter is
designed to isolate players whose long-term contributions are most relevant to teams evaluating
multi-year contracts.

### wRC+ Aging Curve for Top Players

In [5]:
# Specialist model: trained AND scored on the elite cohort.
# Every top-player section pins the training mean as the curve reference.
wrc_top = run_metric(
    WRC, elite_threshold=ELITE_WRC_THRESHOLD, elite_test=True,
    curve_reference="train_mean",
)
print(wrc_top.summary())

plot_aging_curve(wrc_top.curve, WRC, title="wRC+ aging curve — career wRC+ > 110")
plt.show()

wRC+: peak age 28.02  |  peak value 125.1717  |  test MAE 20.1619  (n_train=3929, n_test=411)


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/1437214130.py:10: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Focusing on batters with career wRC+ ≥ 110, the specialist model improves accuracy: test MAE on
strong players drops from 20.56 to 20.16. The peak age is **28**, a full year later than the
all-player peak of 27 — elite hitters do not merely hit better, they hold their level longer
before turning over.

Both the later peak and the accuracy gain point the same way: the decline trajectory for elite
hitters is genuinely different, not just the population curve shifted up, and a model trained on
similarly elite players captures it better.

### WAR Aging Curve for Top players

In [6]:
war_top = run_metric(
    WAR, elite_threshold=ELITE_WAR_THRESHOLD, elite_test=True,
    curve_reference="train_mean",
)
print(war_top.summary())

plot_aging_curve(war_top.curve, WAR, title="WAR aging curve — career WAR > 2.5")
plt.show()

WAR: peak age 27.01  |  peak value 3.9866  |  test MAE 1.6867  (n_train=3581, n_test=376)
  scored PA-weighted; G-weighted (matching the fit) would be 1.6926


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/1331540233.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


The top-player WAR model (career WAR ≥ 2.5) peaks at **age 27** — the same age as the all-player
model — and improves accuracy: test MAE on strong players drops from 1.748 to 1.687.

So for composite value the difference between elite and average players is height, not timing.
That is worth checking against the components, since WAR integrates defense and speed as well as
hitting. If elite players sustain their *hitting* a year longer (peak 28 above) but their
composite WAR still turns at 27, something else in the mix must be pulling the peak back. The
next section examines the defensive aging curve for this same cohort.

#### Explore Denfensive Contribution

In [7]:
# The cohort is defined by career WAR but the metric modelled is Def.
# cohort_metric selects players by IDfg, so the two frames need not align.
def_war_cohort = run_metric(
    DEF, elite_threshold=ELITE_WAR_THRESHOLD, cohort_metric=WAR,
    curve_reference="train_mean",
)
print(def_war_cohort.summary())

plot_aging_curve(def_war_cohort.curve, DEF,
                 title="Def aging curve — career WAR > 2.5 cohort")
plt.show()

Def: peak age 21.99  |  peak value 2.8139  |  test MAE 4.6258  (n_train=3581, n_test=376)
  scored PA-weighted; G-weighted (matching the fit) would be 4.5798


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/706684085.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


The defensive aging curve for the top-WAR cohort (career WAR ≥ 2.5) turns at age 22, the same
place as the all-player population. Elite players do not have a different defensive aging
trajectory; they simply operate from a higher baseline.

The same caution from `GAM.ipynb` applies here: a Def peak at 22 sits barely above the youngest
fitted age and moves by several years under resampling, so it should be read as "declining
throughout" rather than as a located turning point. What matters for the argument is the
direction — defensive value is falling across the whole window in which elite hitting is still
climbing, which is exactly the drag that holds composite WAR's peak at 27 even though these
players' bats do not top out until 28.

To confirm this, the wRC+ aging curve is estimated for the same top-WAR cohort (career WAR ≥ 2.5) rather than the wRC+-filtered cohort used earlier.

In [8]:
wrc_war_cohort = run_metric(
    WRC, elite_threshold=ELITE_WAR_THRESHOLD, cohort_metric=WAR,
    curve_reference="train_mean",
)
print(wrc_war_cohort.summary())

plot_aging_curve(wrc_war_cohort.curve, WRC,
                 title="wRC+ aging curve — career WAR > 2.5 cohort")
plt.show()

wRC+: peak age 28.02  |  peak value 121.9752  |  test MAE 19.8994  (n_train=3581, n_test=376)


/var/folders/8b/6zl3jw494csfbtfj2_zm0tvm0000gn/T/ipykernel_71828/2043535031.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Modelling wRC+ for the top-WAR cohort gives a peak of **28** and the best test MAE of any wRC+
model, 19.90 — better than both the all-player model scored on these rows (20.56) and the
wRC+-filtered specialist (20.16).

The peak matches the wRC+ ≥ 110 cohort exactly, which strengthens rather than weakens the
reading: two differently-defined elite cohorts independently peak a year after the general
population. Their hitting peak at 28 against a composite WAR peak at 27 is the one-year gap
predicted above — the drag from defense and speed, both already past their maxima at this age.